File: week06/15/1_Monday_brackets.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(S):
    # Implement your solution here
    if len(S) == 0:
        return 1    
    if len(S) == 1 or (S[0] in (")", "}", "]")):
        return 0

    bracket = []
    for i in range(len(S)):
        if S[i] in ("[", "{", "("):
            bracket.append(S[i])

        elif S[i] == ")":
            if bracket and bracket[-1] == "(":
                bracket.pop()
            else:
                return 0
        elif S[i] == "}":
            if bracket and bracket[-1] == "{":
                bracket.pop()
            else:
                return 0
        elif S[i] == "]":
            if bracket and bracket[-1] == "[":
                bracket.pop() 
            else:
                return 0
    return 1 if not bracket else 0

File: week06/15/1_Monday_brackets.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

#include <string>

#include <stack>

using namespace std;

int solution(string &S) {
    // Implement your solution here
    stack<char> bracket;
    for (char c : S)
        if (c == '(' || c == '{' || c == '[')
            bracket.push(c);
        
        else if (c == ')')
            if (bracket.top() != '(') return 0;
            else bracket.pop();
        else if (c == '}') 
            if (bracket.top() != '{') return 0;
            else bracket.pop();
        else 
            if (bracket.top() != '[') return 0;
            else bracket.pop();
    
    if (bracket.empty()) return 1;
    else return 0;            
}